### ISSUE TITLE
Not a valid hex SHA1

### ISSUE THREAD

**Issue by CharlesErickT** [Backblaze/B2_Command_Line_Tool#523]
Hello,

I'm getting the following error when running sync
`ERROR:b2.bucket:error when uploading, upload_url was https://pod-000-1050-17.backblaze.com/b2api/v1/b2_upload_file/d16e387f720d6eaa636a0515/c000_v0001050_t0032
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/b2/bucket.py", line 415, in _upload_small_file
    content_type, HEX_DIGITS_AT_END, file_info, hashing_stream
  File "/usr/local/lib/python2.7/dist-packages/b2/raw_api.py", line 497, in upload_file
    return self.b2_http.post_content_return_json(upload_url, headers, data_stream)
  File "/usr/local/lib/python2.7/dist-packages/b2/b2http.py", line 275, in post_content_return_json
    response = _translate_and_retry(do_post, try_count, post_params)
  File "/usr/local/lib/python2.7/dist-packages/b2/b2http.py", line 119, in _translate_and_retry
    return _translate_errors(fcn, post_params)
  File "/usr/local/lib/python2.7/dist-packages/b2/b2http.py", line 55, in _translate_errors
    int(error['status']), error['code'], error['message'], post_params
UnknownError: Unknown error: 400 bad_request Not a valid hex SHA1: [2018-10-15 00:00:55,876] <discover> INF
ERROR:b2.sync.action:an exception occurred in a sync action
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/b2/sync/action.py", line 42, in run
    self.do_action(bucket, reporter)
  File "/usr/local/lib/python2.7/dist-packages/b2/sync/action.py", line 84, in do_action
    progress_listener=SyncFileReporter(reporter)
  File "/usr/local/lib/python2.7/dist-packages/logfury/v0_1/trace_call.py", line 84, in wrapper
    return function(*wrapee_args, **wrapee_kwargs)
  File "/usr/local/lib/python2.7/dist-packages/b2/bucket.py", line 390, in upload
    return f.result()
  File "/usr/local/lib/python2.7/dist-packages/concurrent/futures/_base.py", line 462, in result
    return self.__get_result()
  File "/usr/local/lib/python2.7/dist-packages/concurrent/futures/thread.py", line 63, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python2.7/dist-packages/b2/bucket.py", line 415, in _upload_small_file
    content_type, HEX_DIGITS_AT_END, file_info, hashing_stream
  File "/usr/local/lib/python2.7/dist-packages/b2/raw_api.py", line 497, in upload_file
    return self.b2_http.post_content_return_json(upload_url, headers, data_stream)
  File "/usr/local/lib/python2.7/dist-packages/b2/b2http.py", line 275, in post_content_return_json
    response = _translate_and_retry(do_post, try_count, post_params)
  File "/usr/local/lib/python2.7/dist-packages/b2/b2http.py", line 119, in _translate_and_retry
    return _translate_errors(fcn, post_params)
  File "/usr/local/lib/python2.7/dist-packages/b2/b2http.py", line 55, in _translate_errors
    int(error['status']), error['code'], error['message'], post_params
UnknownError: Unknown error: 400 bad_request Not a valid hex SHA1: [2018-10-15 00:00:55,876] <discover> INF
ERROR:b2.console_tool:ConsoleTool command error
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/b2/console_tool.py", line 1388, in run_command
    return command.run(args)
  File "/usr/local/lib/python2.7/dist-packages/b2/console_tool.py", line 1170, in run
    allow_empty_source=allow_empty_source
  File "/usr/local/lib/python2.7/dist-packages/logfury/v0_1/trace_call.py", line 84, in wrapper
    return function(*wrapee_args, **wrapee_kwargs)
  File "/usr/local/lib/python2.7/dist-packages/b2/sync/sync.py", line 220, in sync_folders
    raise CommandError('sync is incomplete')
CommandError: sync is incomplete
ERROR: sync is incomplete`

Any idea what might cause this ?

Thanks

**Comment by svonohr** [comment]
That's a strange error. The B2 server expected a SHA1 checksum after your file (automatically inserted by B2 CLI) but instead it got `[2018-10-15 00:00:55,876] <discover> INF`. This looks like a part from a log file. Does this file get new data appended while the upload is in progress?

**Comment by ppolewicz** [comment]
We could `seek` to the end of file, `tell` to see where we are, seek to zero and then read the file until the end we recorded, not until the actual end of the file (which may be in a different place already).

I don't think this should be the default behavior - it increases the amount of syscalls and potentially disk operations too, when only the users who try to back up files which are being written during the backup can be affected and need the additional measures to avoid trouble. B2 CLI has been out for a while and this is the first time the issue was reported.

**Comment by bwbeach** [comment]
We already get the file size before uploading, to set the `Content-Length` header.  

The code could just stop reading after that length.  Or report an error that the file is changing?  The SHA1 won't match what was computed at the beginning, so the upload is going to fail anyway.

**Comment by ppolewicz** [comment]
If the file is appended to and we stop reading both for hashing and sending data consistently (at the same offset) then the upload should be successful. I think the issue might be that `StreamWithHash` reads until the end of file, not to the designated offset.

**Comment by bwbeach** [comment]
Would it be more helpful to report a problem if the file has changed since the upload started?  It's possible that data in the middle of the file got rewritten, which could happen if it's an sqlite database, or something like that.

If we decide to attempt the upload even if the file size is bigger, and the checksum doesn't match, there will be an error.  It would be helpful to include "... which may be because the file has changed" in the error message.

**Comment by CharlesErickT** [comment]
> That's a strange error. The B2 server expected a SHA1 checksum after your file (automatically inserted by B2 CLI) but instead it got `[2018-10-15 00:00:55,876] <discover> INF`. This looks like a part from a log file. Does this file get new data appended while the upload is in progress?
> 
> Edit: I managed to reproduce this behavior by appending to a file while being uploaded. What's the expected behavior here? Just uploading as many bytes as were available when the upload started? I think opening all files for exclusive access so nobody can change them is also not feasible.

You are right. The workaround was to rsync the files to another folder and launch the b2 sync from the said folder.

**Comment by 0x3333** [comment]
Would be nice to have the filename that the error occurred in the log file, so debugging is easier.

**Comment by przadka** [comment]
It think this issue can be closed because it is not longer valid. I could not replicate  the error with the current version of the command line tool. Also, looking at the SDK code, the process now takes the initial size of the file into account, as suggested by @ppolewicz earlier.

Current verssion:

```python 
input_stream = ReadingStreamWithProgress(
                            file, progress_listener, length=content_length
                        )
```
vs in the version from when the issue was originally reported:

```python 
input_stream = ReadingStreamWithProgress(file, progress_listener)
```
